ccm_web/client/src/pages/BulkSectionCreate.tsx

```diff
@@ -0,0 +1,534 @@
+    }
+    return invalidations
+  }
+}
+
+// For validating row level issues
+interface SectionsRowInvalidation {
+  message: string
+  rowNumber: number
+  type: InvalidationType
+}
+
+interface SectionRowsValidator {
+  validate: (sectionName: string[]) => SectionsRowInvalidation[]
+}
+
+class DuplicateSectionInFileSectionRowsValidator implements SectionRowsValidator {
+  validate = (sectionNames: string[]): SectionsRowInvalidation[] => {
+    const sortedSectionNames = sectionNames.map(n => { return n.toUpperCase() }).sort((a, b) => { return a.localeCompare(b) })
+    const duplicates: string[] = []
+    for (let i = 1; i < sortedSectionNames.length; ++i) {
+      if (sortedSectionNames[i - 1] === sortedSectionNames[i] && !duplicates.includes(sortedSectionNames[i])) {
+        duplicates.push(sortedSectionNames[i])
+      }
+    }
+    if (duplicates.length === 0) {
+      return []
+    }
+    const invalidations: SectionsRowInvalidation[] = []
+    let i = 1
+    sectionNames.forEach(sectionName => {
+      if (duplicates.includes(sectionName.toUpperCase())) {
+        invalidations.push({ message: 'Duplicate section name: "' + sectionName + '"', rowNumber: i + 1, type: InvalidationType.Error })
+      }
+      ++i
+    })
+    return invalidations
+  }
+}
+
+interface BulkSectionCreateProps extends CCMComponentProps {
+
+}
+
+function BulkSectionCreate (props: BulkSectionCreateProps): JSX.Element {
+  const classes = useStyles()
+  const confirmationClasses = useConfirmationStyles()
+  const rowLevelErrorClasses = useRowLevelErrorStyles()
+  const topLevelClasses = useTopLevelErrorStyles()
+  const apiErrorClasses = useAPIErrorStyles()
+
+  const [isExistingSectionsLoading, setIsExistingSectionsLoading] = useState(true)
+  const [pageState, setPageState] = useState<BulkSectionCreatePageStateData>({ state: BulkSectionCreatePageState.LoadingExistingSectionNames, schemaInvalidation: [], rowInvalidations: [] })
+  const [file, setFile] = useState<File|undefined>(undefined)
+  const [sectionNames, setSectionNames] = useState<string[]>([])
+  const [existingSectionNames, setExistingSectionNames] = useState<string[]|undefined>(undefined)
+
+  const loadCanvasSectionData = (): void => {
+    getCourseSections(props.ltiKey, 'TODO-CourseNumberFromProps?')
+      .then(sections => {
+        setExistingSectionNames(sections.map(s => { return s.toUpperCase() }))
+        setIsExistingSectionsLoading(false)
+      }).catch(() => {
+        setExistingSectionNames(undefined)
+        setPageState({ state: BulkSectionCreatePageState.LoadingExistingSectionNamesFailed, schemaInvalidation: [], rowInvalidations: [] })
+      })
+  }
+
+  // Support for this might be addressed by https://github.com/tl-its-umich-edu/canvas-course-manager-next/issues/74 ?
+  useEffect(() => {
+    loadCanvasSectionData()
+  }, [])
+
+  class DuplicateExistingSectionRowsValidator implements SectionRowsValidator {
```

Comment:
Saw that you moved the validators to separate files. Looks okay to me. I did see that those files are in the `components` directory. Since they aren't React components, I could see moving them into a separate directory one level up, maybe called `validation`? Maybe that's starting to get too far away from where they are used, but I could see that being okay. FWIW, I was originally suggesting that the validators maybe shouldn't be defined in the React component, but they could still be in the same file. I know @pushyamig said differently WRT this file being too large.

Reply:
I didn't like moving them out, but I was trying to react to the comments.  I like moving them into a different directory even more, though I agree they aren't components so it's kind of annoying that way.  Looks like I've managed to come up with a solution that all 3 of us don't like.

Reply:
Okay, maybe some better strategy (that we're all okay with) will become apparent as more validation is created. Thanks